server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/Monitor.java

```diff
@@ -98,54 +127,246 @@ private void monitor(ProcessRef processRef) {
     watcherThread.start();
     watcherThreads.add(watcherThread);
 
-    processes.add(processRef);
-
     // wait for process to be ready (accept requests or so on)
     processRef.waitForReady();
 
-    LoggerFactory.getLogger(getClass()).info(String.format("%s is up", processRef));
+    LOG.info("{} is up", processRef);
+  }
+
+  private void stopIfAnyProcessDidNotStart() {
+    if (!lifecycle.tryToMoveTo(State.STARTED)) {
+      // stopping or stopped during startup, for instance :
+      // 1. A is started
+      // 2. B starts
+      // 3. A crashes while B is starting
+      // 4. if B was not monitored during Terminator execution, then it's an alive orphan
+      stop();
+      throw new IllegalStateException("Stopped during startup");
+    }
   }
 
   /**
    * Blocks until all processes are terminated
    */
   public void awaitTermination() {
+    while (awaitChildProcessesTermination()) {
+      trace("await termination of restartor...");
+      awaitTermination(restartor);
+    }
+    cleanAfterTermination();
+  }
+
+  boolean waitForOneRestart() {
+    boolean restartRequested = awaitChildProcessesTermination();
+    trace("finished waiting, restartRequested={}", restartRequested);
+    if (restartRequested) {
+      trace("awaitTermination restartor={}", restartor);
+      awaitTermination(restartor);
+    }
+    return restartRequested;
+  }
+
+  private boolean awaitChildProcessesTermination() {
+    trace("await termination of child processes...");
+    List<WatcherThread> watcherThreads = new ArrayList<>(this.watcherThreads);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Rename "watcherThreads" which hides the field declared at line 52. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)